cli/src/repl.rs

```diff
@@ -1,47 +1,230 @@
+    let mut editor: REPLEditor =
+        DefaultEditor::new().expect("unable to instantiate terminal editor");
+    editor.set_color_mode(ColorMode::Enabled);
+    editor.set_history_ignore_dups(true).unwrap();
+    editor.set_history_ignore_space(true);
+    editor.bind_sequence(
+        Event::KeySeq(vec![KeyEvent(KeyCode::Char('u'), Modifiers::ALT)]),
+        EventHandler::from(Cmd::Undo(1)),
+    );
+
+    let mut relations = Relations::default();
+    let mut imports = Imports::default();
+    let mut importer = REPLImporter::default();
+
+    loop {
+        let mut engine = Engine::default();
+
+        let source = parse_input(&mut editor);
+        handle_source(
+            source,
+            &config,
+            &mut engine,
+            &mut importer,
+            &mut imports,
+            &mut relations,
+        );
+    }
+}
+
+fn strip_indent(line: &mut String) -> String {
+    let mut indent = String::new();
+    let line_clone = line.to_string();
+    let chars = line_clone.chars();
+    for c in chars {
+        if !c.is_whitespace() {
+            return indent;
+        }
+        line.remove(0);
+        indent.push(c);
     }
+    indent
 }
 
 /// Parses stdin until the user's input forms a source code with no unclosed delimiters
 /// and return the source.
-fn parse_input() -> Option<OwnedSource> {
-    let stdin = io::stdin();
-    let lines = stdin.lock().lines();
+fn parse_input(editor: &mut REPLEditor) -> OwnedSource {
     let mut content = String::new();
-    print_flush!("=> ");
-    for line in lines {
-        let line = line.expect("couldn't read line from stdin");
+    let mut prompt_prefix = "=> ".to_string();
+    let mut indent_prefix = String::new();
+    loop {
+        let line = editor.readline_with_initial(&prompt_prefix, (&indent_prefix, ""));
+        let mut line = match line {
+            Ok(line) => line,
+            Err(ReadlineError::Interrupted) | Err(ReadlineError::Eof) => exit(1),
+            e => e.expect("error when reading next line from editor"),
+        };
+
+        indent_prefix = strip_indent(&mut line);
+
         content.push_str(&line);
         if line.ends_with('\\') {
             content.push('\n');
-            print_flush!("-> ");
+            prompt_prefix = "-> ".to_string();
             continue;
         }
 
         let source = Source::new(&content, "stdin");
         let report = parse(source);
-        if !report.stack_ended {
+        if let Some(last) = report.delimiter_stack.last() {
             content.push('\n');
-            print_flush!("-> ");
+            prompt_prefix = format!("{}> ", last.str().unwrap());
             continue; // Silently ignore incomplete input
         }
 
-        return Some(OwnedSource::new(
-            source.source.to_string(),
-            source.name.to_string(),
-        ));
+        editor
+            .add_history_entry(source.source.to_string())
+            .expect("terminal has no history");
```

Comment:
```suggestion
            .ok();
```
Don't panic just because the line cannot be appended to the history.